### PR TITLE
[contacts] implement picking contact via native UI

### PR DESCRIPTION
**Pull request by fobos531**
# Why

https://expo.canny.io/feature-requests/p/use-native-contact-pickers

# How

https://github.com/expo/expo/issues/2645

# Test Plan

Added a super basic example in ContactsScreen which showcases the usage of this. Example:

https://github.com/expo/expo/assets/19533289/9dbab110-f987-44b7-941d-c490b2558bb8

https://github.com/expo/expo/assets/19533289/bec5483b-c2a7-4466-84cf-b3431473d39b

# Checklist

- [X] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [X] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [X] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

**Notes**:

- as this an initial implementation of this feature, I'm looking for any and all feedback that you may have regarding the actual API (method names, function signatures, additional options etc.)

### CODE REVIEW

**Comment by fobos531** [comment]
@alanjhughes Even though Evan is the codeowner for this package, you may want to take a look at it since you were doing the migration of the Contacts module to the Sweet API.

File: packages/expo-contacts/src/Contacts.ts
Comment (fobos531) [inline packages/expo-contacts/src/Contacts.ts:948]: @amandeepmittal On Android, for this to work, contact permissions need to be granted first, but on iOS this is not necessary (permissions are handled automatically by iOS). What is the proper way/syntax to highlight this in the docs?

Comment (amandeepmittal) [reply packages/expo-contacts/src/Contacts.ts:948]: Hey @fobos531, you can describe this in the method definition both for Android and iOS. When adding the description, for Android, you can link to [`requestPermissionAsync()`](https://docs.expo.dev/versions/latest/sdk/contacts/#contactsrequestpermissionsasync) since that is used to ask for user permission. This will help folks to be aware that they need to call this function in their app code for Android.

**Comment by fobos531** [comment]
Hey @brentvatne or @alanjhughes Do either of you have the bandwith to give a first pass review of this PR? Would love to incorporate your feedback if you have any.

I'm trying to test my changes on android, but I'm currently running into build errors, something related to the `expo-modules-core` - do you have any pointers on what could be causing this? See the build error log: https://gist.github.com/fobos531/9d90f074e0050472e9300fead357200c

**Comment by alanjhughes** [comment]
Hey @fobos531 sorry for the delay. I will have a look today or tomorrow. Regarding the build issues, try deleting the .cxx and .build directories for-expo modules-core. Run `./gradlew clean` and then `./gradlew assembleDebug` in `bare-expo/android`.

**Comment by fobos531** [comment]
> Hey @fobos531 sorry for the delay. I will have a look today or tomorrow. Regarding the build issues, try deleting the .cxx and .build directories in expo-modules-core. Run `./gradlew clean` and then `./gradlew assembleDebug` in `bare-expo/android`.

That worked great, thanks! I've tested and verified that Android works just fine as well with the changes proposed so far. I've updated the OP to reflect this.

File: packages/expo-contacts/ios/ContactPickerControllerDelegate.swift
Comment (alanjhughes) [inline packages/expo-contacts/ios/ContactPickerControllerDelegate.swift:14]:
```suggestion
```

File: packages/expo-contacts/ios/ContactPickerControllerDelegate.swift
Comment (alanjhughes) [inline packages/expo-contacts/ios/ContactPickerControllerDelegate.swift:22]: Can you move the initializer above the functions please

File: packages/expo-contacts/ios/ContactsModule.swift
Comment (alanjhughes) [inline packages/expo-contacts/ios/ContactsModule.swift:163]: Don't think we need the context here. It's just storing the promise so we could store it on the module directly like `contactManipulationPromise`

File: packages/expo-contacts/ios/ContactsModule.swift
Comment (alanjhughes) [reply packages/expo-contacts/ios/ContactsModule.swift:341]: `ContactSerializationException`? There's not much else we can do in this case

File: packages/expo-contacts/ios/ContactsModule.swift
Comment (alanjhughes) [inline packages/expo-contacts/ios/ContactsModule.swift:349]:
```suggestion
    currentContactPickingContext?.promise.resolve()
```

File: packages/expo-contacts/ios/ContactsModule.swift
Comment (alanjhughes) [inline packages/expo-contacts/ios/ContactsModule.swift:10]:
```suggestion
```
See comment about storing the promise directly

File: packages/expo-contacts/android/src/main/java/expo/modules/contacts/ContactsModule.kt
Comment (alanjhughes) [inline packages/expo-contacts/android/src/main/java/expo/modules/contacts/ContactsModule.kt:280]:
```suggestion
          pendingPromise.resolve()
```

**Review by alanjhughes — COMMENTED** [review]
Looks good @fobos531 👍 Just a few comments

**Comment by fobos531** [comment]
> Looks good @fobos531 👍 Just a few comments

Done @alanjhughes . Let me know if you require any additional changes from my end.

**Comment by fobos531** [comment]
Not sure why check-packages workflow is failing 🤔 I rebuilt JS but git doesn't detect any changes that would be made...

**Comment by alanjhughes** [comment]
@fobos531 using `throws` works too 👍 regarding `check-packages`, are you using node 20? If not, can you use it and rebuild the js again?

**Comment by fobos531** [comment]
> @fobos531 using `throws` works too 👍 regarding `check-packages`, are you using node 20? If not, can you use it and rebuild the js again?

Yes, I did just that and I confirmed that I am using node 20. Moreover, it seems to fail on `expo-updates` as well, but I wasn't making any changes to that package whatsoever. So I guess the action is just acting up?

**Review by alanjhughes — APPROVED** [review]

File: packages/expo-contacts/ios/ContactPickerControllerDelegate.swift
Comment (alanjhughes) [inline packages/expo-contacts/ios/ContactPickerControllerDelegate.swift:19]: Looking at this now, we'll need to deal with this error and clear the promise after. Otherwise it will fail, and attempting to try again will also fail because the promise won't be nil

**Review by alanjhughes — CHANGES_REQUESTED** [review]

**Review by alanjhughes — APPROVED** [review]

**Comment by alanjhughes** [comment]
@fobos531 you can ignore the updates issue, I've addressed it in #27891. Just need to rebuild the contacts js and we should be good to merge

**Comment by fobos531** [comment]
> @fobos531 you can ignore the updates issue, I've addressed it in #27891. Just need to rebuild the contacts js and we should be good to merge

@alanjhughes Seems like `expo-contacts` is passing now. `expo-dev-launcher` is failing, but seems to be unrelated

**Comment by fobos531** [comment]
Hey @alanjhughes , is there anything else left that would be blocking this PR?

File: packages/expo-contacts/src/Contacts.ts
Comment (amandeepmittal) [inline packages/expo-contacts/src/Contacts.ts:948]:
```suggestion
 * Presents a native contact picker to select a single contact from the system. On Android, the `READ_CONTACTS` permission is required. You can
 * obtain this permission by calling the [Contacts.requestPermissionsAsync()](#contactsrequestpermissionsasync) method. On iOS, no permissions are 
 * required to use this method. 
```

Let's use the correct platform order [as per our writing style guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md#referencing-android-ios-and-web).

**Comment by alanjhughes** [comment]
Hey @fobos531, sorry for the delay. All looks good now. Just waiting on CI. Thanks for the contribution!

**Comment by fzf** [comment]
I want to say this is amazing and I love it, I was wondering if there was a reason the ability to pass displayedKeys into the iOS was not implemented (it will allow you to allow people to select a phone number if there are multiple for instance) is there no such mechanism in Android and so this is to maintain consistency? Just wanted to get some info before I dived deeper into it.

**Comment by fobos531** [comment]
> displayedKeys

Hey @fzf. I wasn't aware of this at all, so thats the primary reason why I didn't add it. Hope this helps

**Pull request merged**
